kubectl-openebs/cli/command/get/volume_list.go

```diff
@@ -44,65 +47,102 @@ func NewCmdGetVolume() *cobra.Command {
+func RunVolumesList(casType string, vols []string) error {
 	k8sClient, err := client.NewK8sClient("")
-	util.CheckErr(err, util.Fatal)
-	if openebsNs == "" {
-		nsFromCli, err := k8sClient.GetOpenEBSNamespace(util.CstorCasType)
-		if err != nil {
-			//return errors.Wrap(err, "Error determining the openebs namespace, please specify using \"--openebs-namespace\" flag")
-			return errors.New("no cstor volumes found in the cluster")
-		}
-		k8sClient.Ns = nsFromCli
+	if err != nil {
+		return err
 	}
-	var cvols *v1.CStorVolumeList
+	// 1. Fetch all or required PVs
+	var pvList *corev1.PersistentVolumeList
 	if len(vols) == 0 {
-		cvols, err = k8sClient.GetcStorVolumes()
+		pvList, err = k8sClient.GetPVs()
 	} else {
-		cvols, err = k8sClient.GetcStorVolumesByNames(vols)
+		pvList, err = k8sClient.GetPVbyName(vols)
 	}
 	if err != nil {
-		return errors.Wrap(err, "error listing volumes")
+		return err
 	}
-	pvols, err := k8sClient.GetCStorVolumeInfoMap("")
-	if err != nil {
-		return errors.Wrap(err, "failed to execute volume info command")
+	// 2. Fetch all relevant volume CRs without worrying about openebsNS
+	var (
+		jvMap  map[string]v1alpha1.JivaVolume
+		cvMap  map[string]v1.CStorVolume
+		cvaMap map[string]v1.CStorVolumeAttachment
+	)
+	if casType == "" {
+		// fetch all
+		jvMap, _ = k8sClient.GetJivaVolumeMap()
+		cvMap, _ = k8sClient.GetCStorVolumeMap()
+		cvaMap, _ = k8sClient.GetCStorVolumeAttachmentMap()
+	} else if casType == util.JivaCasType {
+		jvMap, _ = k8sClient.GetJivaVolumeMap()
+	} else if casType == util.CstorCasType {
+		cvMap, _ = k8sClient.GetCStorVolumeMap()
+		cvaMap, _ = k8sClient.GetCStorVolumeAttachmentMap()
 	}
-	// tally status of cvols to pvols
-	// give output according to volume status
 	var rows []metav1.TableRow
-	for _, item := range cvols.Items {
-		storageClass := util.NotAttached
-		attachementStatus := util.NotAttached
-		accessMode := util.NotAttached
-		node := util.NotAttached
-		if val, ok := pvols[item.ObjectMeta.Name]; ok {
-			storageClass = val.StorageClass
-			attachementStatus = val.AttachementStatus
-			accessMode = val.AccessMode
-			node = val.Node
+	// 3. Show the required ones
+	for _, pv := range pvList.Items {
+		name := pv.Name
+		capacity := pv.Spec.Capacity.Storage()
+		sc := pv.Spec.StorageClassName
+		attached := pv.Status.Phase
+		var attachedNode, storageVersion, customStatus, ns string
+		// TODO: Estimate the cas-type and decide to print it out
+		// Should all AccessModes be shown in a csv format, or the highest be displayed ROO < RWO < RWX?
+		if pv.Spec.CSI != nil {
+			// 2. For eligible PVs fetch the custom-resource to add more info
+			if pv.Spec.CSI.Driver == util.CStorCSIDriver && (casType == util.CstorCasType || casType == "") {
+				// For all CSI CStor PV, there exist a CV
+				cv, ok := cvMap[pv.Name]
+				if !ok {
+					// condition not possible
+					return fmt.Errorf("couldn't find cv %s", pv.Name)
```

Comment:
We should not terminate here. There can be a case where suppose 10 cstor pvs exist and out of which 9 have a cv and 1 doesnot have. If we return here the 9 pvs having cv would not get listed.